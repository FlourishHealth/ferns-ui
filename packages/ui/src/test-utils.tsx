import {render} from "@testing-library/react-native";
import React from "react";
import {ThemeProvider} from "./Theme";

export const renderWithTheme = (ui: React.ReactElement) => {
  return render(<ThemeProvider>{ui}</ThemeProvider>);
};

export const createCommonMocks = () => ({
  onChange: jest.fn(),
  onFocus: jest.fn(),
  onBlur: jest.fn(),
  onEnter: jest.fn(),
  onSubmitEditing: jest.fn(),
  onIconClick: jest.fn(),
});

export const setupComponentTest = () => {
  jest.useFakeTimers();
  return createCommonMocks();
};

export const teardownComponentTest = () => {
  jest.useRealTimers();
  jest.clearAllMocks();
};
