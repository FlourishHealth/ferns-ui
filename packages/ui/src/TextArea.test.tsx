import {userEvent} from "@testing-library/react-native";
import React from "react";

import {TextArea} from "./TextArea";
import {renderWithTheme} from "./test-utils";

describe("TextArea", () => {
  let mockOnChange: jest.Mock;

  beforeEach(() => {
    mockOnChange = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("basic functionality", () => {
    it("should render as multiline text field", () => {
      const {getByDisplayValue} = renderWithTheme(
        <TextArea value="test content" onChange={mockOnChange} />
      );
      
      const input = getByDisplayValue("test content");
      expect(input.props.multiline).toBe(true);
      expect(input.props.value).toBe("test content");
    });

    it("should render with title", () => {
      const {getByText} = renderWithTheme(
        <TextArea title="Description" value="" onChange={mockOnChange} />
      );
      
      expect(getByText("Description")).toBeTruthy();
    });

    it("should render with placeholder", () => {
      const {getByPlaceholderText} = renderWithTheme(
        <TextArea placeholder="Enter description" value="" onChange={mockOnChange} />
      );
      
      expect(getByPlaceholderText("Enter description")).toBeTruthy();
    });

    it("should render helper text", () => {
      const {getByText} = renderWithTheme(
        <TextArea helperText="Maximum 500 characters" value="" onChange={mockOnChange} />
      );
      
      expect(getByText("Maximum 500 characters")).toBeTruthy();
    });

    it("should render error text", () => {
      const {getByText} = renderWithTheme(
        <TextArea errorText="This field is required" value="" onChange={mockOnChange} />
      );
      
      expect(getByText("This field is required")).toBeTruthy();
    });

    it("should support grow behavior", () => {
      const {getByDisplayValue} = renderWithTheme(
        <TextArea grow value="" onChange={mockOnChange} />
      );
      
      const input = getByDisplayValue("");
      expect(input.props.multiline).toBe(true);
    });

    it("should be disabled when disabled prop is true", () => {
      const {getByDisplayValue} = renderWithTheme(
        <TextArea disabled value="test" onChange={mockOnChange} />
      );
      
      const input = getByDisplayValue("test");
      expect(input.props.readOnly).toBe(true);
    });

    it("should always have text type", () => {
      const {getByDisplayValue} = renderWithTheme(
        <TextArea value="" onChange={mockOnChange} />
      );
      
      const input = getByDisplayValue("");
      expect(input.props.keyboardType).toBe("default");
    });
  });

  describe("user interactions", () => {
    it("should call onChange when text is entered", async () => {
      const user = userEvent.setup();
      const {getByDisplayValue} = renderWithTheme(
        <TextArea value="" onChange={mockOnChange} />
      );

      const input = getByDisplayValue("");
      await user.type(input, "hello world");

      expect(mockOnChange).toHaveBeenCalled();
      expect(mockOnChange.mock.calls.length).toBeGreaterThan(0);
    });

    it("should handle multiline text input", async () => {
      const user = userEvent.setup();
      const multilineText = "Line 1\nLine 2\nLine 3";
      const {getByDisplayValue} = renderWithTheme(
        <TextArea value="" onChange={mockOnChange} />
      );

      const input = getByDisplayValue("");
      await user.type(input, multilineText);

      expect(mockOnChange).toHaveBeenCalled();
      expect(mockOnChange.mock.calls.length).toBeGreaterThan(0);
    });
  });

  describe("accessibility", () => {
    it("should have correct accessibility properties", () => {
      const {getByDisplayValue} = renderWithTheme(
        <TextArea value="" onChange={mockOnChange} />
      );
      
      const input = getByDisplayValue("");
      expect(input.props.accessibilityHint).toBe("Enter text here");
      expect(input.props["aria-label"]).toBe("Text input field");
    });

    it("should indicate disabled state in accessibility", () => {
      const {getByDisplayValue} = renderWithTheme(
        <TextArea disabled value="" onChange={mockOnChange} />
      );
      
      const input = getByDisplayValue("");
      expect(input.props.accessibilityState.disabled).toBe(true);
    });
  });

  describe("edge cases", () => {
    it("should handle empty value", () => {
      const {getByDisplayValue} = renderWithTheme(
        <TextArea value="" onChange={mockOnChange} />
      );
      
      const input = getByDisplayValue("");
      expect(input.props.value).toBe("");
    });

    it("should handle undefined value", () => {
      const {root} = renderWithTheme(
        <TextArea value={undefined} onChange={mockOnChange} />
      );
      
      expect(root).toBeTruthy();
    });

    it("should handle long text values", () => {
      const longText = "Lorem ipsum dolor sit amet, consectetur adipiscing elit. ".repeat(50);
      const {getByDisplayValue} = renderWithTheme(
        <TextArea value={longText} onChange={mockOnChange} />
      );
      
      const input = getByDisplayValue(longText);
      expect(input.props.value).toBe(longText);
    });

    it("should handle text with line breaks", () => {
      const textWithBreaks = "First line\nSecond line\n\nFourth line";
      const {getByDisplayValue} = renderWithTheme(
        <TextArea value={textWithBreaks} onChange={mockOnChange} />
      );
      
      const input = getByDisplayValue(textWithBreaks);
      expect(input.props.value).toBe(textWithBreaks);
    });
  });

  describe("props inheritance", () => {
    it("should inherit all TextField props except multiline and type", () => {
      const mockOnFocus = jest.fn();
      const mockOnBlur = jest.fn();
      
      const {getByDisplayValue} = renderWithTheme(
        <TextArea 
          value="test" 
          onChange={mockOnChange}
          onFocus={mockOnFocus}
          onBlur={mockOnBlur}
          placeholder="Test placeholder"
          disabled={false}
          rows={5}
        />
      );
      
      const input = getByDisplayValue("test");
      expect(input.props.numberOfLines).toBe(5);
      expect(input.props.onFocus).toBeTruthy();
      expect(input.props.onBlur).toBeTruthy();
    });

    it("should support inputRef", () => {
      const mockInputRef = jest.fn();
      renderWithTheme(
        <TextArea inputRef={mockInputRef} value="" onChange={mockOnChange} />
      );
      
      expect(mockInputRef).toHaveBeenCalled();
    });
  });
});
