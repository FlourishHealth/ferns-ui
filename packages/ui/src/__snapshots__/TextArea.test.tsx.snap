// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea snapshots should match snapshot when disabled 1`] = `
<View
  style={
    {
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "color": "#1C1C1C",
        "fontSize": 14,
        "fontWeight": 600,
        "lineHeight": 22.4,
      }
    }
  >
    Disabled TextArea
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#D9D9D9",
        "borderColor": "#4E4E4E",
        "borderRadius": 4,
        "borderWidth": 1,
        "flexDirection": "row",
        "overflow": "hidden",
        "paddingHorizontal": 12,
        "paddingVertical": 8,
      }
    }
  >
    <TextInput
      accessibilityHint="Enter text here"
      accessibilityState={
        {
          "disabled": true,
        }
      }
      aria-label="Text input field"
      autoCapitalize="sentences"
      autoCorrect={true}
      blurOnSubmit={true}
      keyboardType="default"
      multiline={true}
      numberOfLines={1}
      onBlur={[Function]}
      onChangeText={[MockFunction]}
      onContentSizeChange={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholderTextColor="#686868"
      readOnly={true}
      secureTextEntry={false}
      style={
        {
          "color": "#1C1C1C",
          "flex": 1,
          "fontFamily": "text",
          "fontSize": 16,
          "gap": 10,
          "height": 40,
          "paddingVertical": 0,
          "width": "100%",
        }
      }
      textContentType="none"
      underlineColorAndroid="transparent"
      value="disabled content"
    />
  </View>
</View>
`;

exports[`TextArea snapshots should match snapshot with all props 1`] = `
<View
  style={
    {
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "color": "#1C1C1C",
        "fontSize": 14,
        "fontWeight": 600,
        "lineHeight": 22.4,
      }
    }
  >
    Description
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#9A9A9A",
        "borderRadius": 4,
        "borderWidth": 1,
        "flexDirection": "row",
        "overflow": "hidden",
        "paddingHorizontal": 12,
        "paddingVertical": 8,
      }
    }
  >
    <TextInput
      accessibilityHint="Enter text here"
      accessibilityState={
        {
          "disabled": false,
        }
      }
      aria-label="Text input field"
      autoCapitalize="sentences"
      autoCorrect={true}
      blurOnSubmit={true}
      keyboardType="default"
      multiline={true}
      numberOfLines={5}
      onBlur={[Function]}
      onChangeText={[MockFunction]}
      onContentSizeChange={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Enter description"
      placeholderTextColor="#686868"
      readOnly={false}
      secureTextEntry={false}
      style={
        {
          "color": "#1C1C1C",
          "flex": 1,
          "fontFamily": "text",
          "fontSize": 16,
          "gap": 10,
          "height": 200,
          "paddingVertical": 0,
          "width": "100%",
        }
      }
      textContentType="none"
      underlineColorAndroid="transparent"
      value="test content"
    />
  </View>
  <View
    style={
      {
        "marginTop": 2,
      }
    }
  >
    <Text
      style={
        {
          "color": "#1C1C1C",
          "fontSize": 12,
          "lineHeight": 16,
        }
      }
    >
      Maximum 500 characters
    </Text>
  </View>
</View>
`;

exports[`TextArea snapshots should match snapshot with default props 1`] = `
<View
  style={
    {
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#9A9A9A",
        "borderRadius": 4,
        "borderWidth": 1,
        "flexDirection": "row",
        "overflow": "hidden",
        "paddingHorizontal": 12,
        "paddingVertical": 8,
      }
    }
  >
    <TextInput
      accessibilityHint="Enter text here"
      accessibilityState={
        {
          "disabled": undefined,
        }
      }
      aria-label="Text input field"
      autoCapitalize="sentences"
      autoCorrect={true}
      blurOnSubmit={true}
      keyboardType="default"
      multiline={true}
      numberOfLines={1}
      onBlur={[Function]}
      onChangeText={[MockFunction]}
      onContentSizeChange={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholderTextColor="#686868"
      secureTextEntry={false}
      style={
        {
          "color": "#1C1C1C",
          "flex": 1,
          "fontFamily": "text",
          "fontSize": 16,
          "gap": 10,
          "height": 40,
          "paddingVertical": 0,
          "width": "100%",
        }
      }
      textContentType="none"
      underlineColorAndroid="transparent"
      value="test content"
    />
  </View>
</View>
`;

exports[`TextArea snapshots should match snapshot with error state 1`] = `
<View
  style={
    {
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "color": "#1C1C1C",
        "fontSize": 14,
        "fontWeight": 600,
        "lineHeight": 22.4,
      }
    }
  >
    Error TextArea
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginVertical": 2,
      }
    }
  >
    <View
      style={
        {
          "marginLeft": 4,
        }
      }
    >
      <Text
        style={
          {
            "color": "#BD1111",
            "fontSize": 12,
          }
        }
      >
        This field is required
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#D33232",
        "borderRadius": 4,
        "borderWidth": 1,
        "flexDirection": "row",
        "overflow": "hidden",
        "paddingHorizontal": 12,
        "paddingVertical": 8,
      }
    }
  >
    <TextInput
      accessibilityHint="Enter text here"
      accessibilityState={
        {
          "disabled": undefined,
        }
      }
      aria-label="Text input field"
      autoCapitalize="sentences"
      autoCorrect={true}
      blurOnSubmit={true}
      keyboardType="default"
      multiline={true}
      numberOfLines={1}
      onBlur={[Function]}
      onChangeText={[MockFunction]}
      onContentSizeChange={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholderTextColor="#686868"
      secureTextEntry={false}
      style={
        {
          "color": "#1C1C1C",
          "flex": 1,
          "fontFamily": "text",
          "fontSize": 16,
          "gap": 10,
          "height": 40,
          "paddingVertical": 0,
          "width": "100%",
        }
      }
      textContentType="none"
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
</View>
`;

exports[`TextArea snapshots should match snapshot with multiline content 1`] = `
<View
  style={
    {
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "color": "#1C1C1C",
        "fontSize": 14,
        "fontWeight": 600,
        "lineHeight": 22.4,
      }
    }
  >
    Multiline Content
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#9A9A9A",
        "borderRadius": 4,
        "borderWidth": 1,
        "flexDirection": "row",
        "overflow": "hidden",
        "paddingHorizontal": 12,
        "paddingVertical": 8,
      }
    }
  >
    <TextInput
      accessibilityHint="Enter text here"
      accessibilityState={
        {
          "disabled": undefined,
        }
      }
      aria-label="Text input field"
      autoCapitalize="sentences"
      autoCorrect={true}
      blurOnSubmit={true}
      keyboardType="default"
      multiline={true}
      numberOfLines={4}
      onBlur={[Function]}
      onChangeText={[MockFunction]}
      onContentSizeChange={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholderTextColor="#686868"
      secureTextEntry={false}
      style={
        {
          "color": "#1C1C1C",
          "flex": 1,
          "fontFamily": "text",
          "fontSize": 16,
          "gap": 10,
          "height": 160,
          "paddingVertical": 0,
          "width": "100%",
        }
      }
      textContentType="none"
      underlineColorAndroid="transparent"
      value="Line 1\\nLine 2\\nLine 3"
    />
  </View>
</View>
`;
